### ISSUE TITLE
trace: Support setting event parent events explicitly

### ISSUE THREAD

**Issue by hawkw** [tokio-rs/tokio#1100]
## Motivation

Currently, the `tokio-trace` span macros allow setting the parent span explicitly as a parameter of the macro, as well as contextually. However, events created via the event macros _always_ have their parents determined contextually. It would be nice to allow setting explicit parents for events as well (as @jonhoo mentioned in #1065).

## Implementation

Making this change will require changes in both `tokio-trace` and `tokio-trace-core`. I believe we can do this somewhat like so:
1. In the `tokio-trace-core` crate, move the [`Parent`](https://github.com/tokio-rs/tokio/blob/b2c53987d9af8cdaa6ba49a084bb4363f86bd295/tokio-trace/tokio-trace-core/src/span.rs#L33-L41) type out of the `span` module, and make it `pub(crate)`, so that it can be used in the `event` module as well.
2. Add a `parent: Parent` field to the [`Event`](https://github.com/tokio-rs/tokio/blob/b2c53987d9af8cdaa6ba49a084bb4363f86bd295/tokio-trace/tokio-trace-core/src/event.rs#L21-L24) type.
3. Add a new variant<sup>1</sup> of the [`Event::dispatch`](https://github.com/tokio-rs/tokio/blob/b2c53987d9af8cdaa6ba49a084bb4363f86bd295/tokio-trace/tokio-trace-core/src/event.rs#L30-L35) function to construct and observe an event with an explicit parent, similar to [`span::Attributes::child_of`](https://github.com/tokio-rs/tokio/blob/b2c53987d9af8cdaa6ba49a084bb4363f86bd295/tokio-trace/tokio-trace-core/src/span.rs#L86-L98). I'm not sure whether or not we also want to allow events to be created _without_ parents (similar to `span::Attributes::new_root`), but if so, we can add an explicit no parent function as well.
4. Add the [parent accessor methods](https://github.com/tokio-rs/tokio/blob/b2c53987d9af8cdaa6ba49a084bb4363f86bd295/tokio-trace/tokio-trace-core/src/span.rs#L111-L142) from `span::Attributes` to `Event` as well. We could consider factoring out the logic to the `Parent` type and have `Event` and `Attributes` call those functions on `Parent`, in order to reduce repetition, but the implementation is simple enough that I'm not sure if that's strictly necessary.
5. In the `tokio-trace` crate, add match arms to the [`event!`](https://github.com/tokio-rs/tokio/blob/b2c53987d9af8cdaa6ba49a084bb4363f86bd295/tokio-trace/src/macros.rs#L538-L699) macro that accept a `parent:` field and construct new events with parents. These will look similar to [this match arm](https://github.com/tokio-rs/tokio/blob/b2c53987d9af8cdaa6ba49a084bb4363f86bd295/tokio-trace/src/macros.rs#L132-L158) from the `span!` macro, but changed to [construct events](https://github.com/tokio-rs/tokio/blob/b2c53987d9af8cdaa6ba49a084bb4363f86bd295/tokio-trace/src/macros.rs#L631-L661). 
6. Add similar match arms to the [`error!`, `warn!`, `info!`, `debug!`, and `trace!` macros](https://github.com/tokio-rs/tokio/blob/b2c53987d9af8cdaa6ba49a084bb4363f86bd295/tokio-trace/src/macros.rs#L701-L1007) that expand to the appropriate invocations of the `event!` macro. The similarly-named [leveled span macros](https://github.com/tokio-rs/tokio/blob/b2c53987d9af8cdaa6ba49a084bb4363f86bd295/tokio-trace/src/macros.rs#L234-L536) can be used as a guide.

This should not require a breaking change.

---
<sup>1</sup>We will want to create a new function rather than add a new parameter to the existing `Event::dispatch` function, as adding a new parameter is a breaking change in `tokio-trace-core`.

**Comment by hawkw** [comment]
I've tried to provide a thorough description of how I would recommend making this change, so it should be appropriate for a new contributor. With that said, if anyone's interested in working on this one and anything I said above is confusing or unclear, please let me know and I'm happy to explain further.

**Comment by zaharidichev** [comment]
@hawkw May be wrong on that, but it seems to me that in order to implement `Event::child_of(...)` the same way as `Span::child_of` ,the `AsId` type needs to be moved from tokio-trace to tokio-trace-core in order to be visible there. Correct me if I am wrong :)

**Comment by hawkw** [comment]
@zaharidichev that's a good point, thanks --- I think I overlooked that. I'd actually like to look into removing the `AsId` trait in the upcoming release, as I think it can be replaced with `Into` implementations. So, that may be the best solution.

**Comment by zaharidichev** [comment]
Wouldn’t that introduce a breaking change though ?

**Comment by hawkw** [comment]
Yes. There are other unavoidable breaking changes scheduled for a `tokio-trace` 0.2 release, though, so now would be a good time to make that change as well.

**Comment by hawkw** [comment]
@zaharidichev see also #1102 :)

**Comment by zaharidichev** [comment]
Great, I already started on this one and will rebase once #1102 is merged. Will be able to pick it back up in the following days though. Someone might beat me to it by then :)
